plugins/catalog-backend/src/service/CatalogBuilder.ts

```diff
@@ -0,0 +1,523 @@
+  makeValidator,
+  NoForeignRootFieldsEntityPolicy,
+  ReservedFieldsEntityPolicy,
+  SchemaValidEntityPolicy,
+  templateEntityV1alpha1Policy,
+  userEntityV1alpha1Policy,
+  Validators,
+} from '@backstage/catalog-model';
+import { Config } from '@backstage/config';
+import lodash from 'lodash';
+import { Logger } from 'winston';
+import {
+  DatabaseEntitiesCatalog,
+  DatabaseLocationsCatalog,
+  EntitiesCatalog,
+  LocationsCatalog,
+} from '../catalog';
+import { DatabaseManager } from '../database';
+import {
+  AnnotateLocationEntityProcessor,
+  AzureApiReaderProcessor,
+  BitbucketApiReaderProcessor,
+  CodeOwnersProcessor,
+  EntityPolicyProcessor,
+  FileReaderProcessor,
+  GithubOrgReaderProcessor,
+  GithubReaderProcessor,
+  GitlabApiReaderProcessor,
+  GitlabReaderProcessor,
+  HigherOrderOperation,
+  HigherOrderOperations,
+  LocationProcessor,
+  LocationReaders,
+  LocationRefProcessor,
+  PlaceholderProcessor,
+  PlaceholderResolver,
+  StaticLocationProcessor,
+  UrlReaderProcessor,
+  YamlProcessor,
+} from '../ingestion';
+import { CatalogRulesEnforcer } from '../ingestion/CatalogRules';
+import { LdapOrgReaderProcessor } from '../ingestion/processors/LdapOrgReaderProcessor';
+import {
+  jsonPlaceholderResolver,
+  textPlaceholderResolver,
+  yamlPlaceholderResolver,
+} from '../ingestion/processors/PlaceholderProcessor';
+
+export type CatalogEnvironment = {
+  logger: Logger;
+  database: PluginDatabaseManager;
+  config: Config;
+  reader: UrlReader;
+};
+
+/**
+ * A builder that helps wire up all of the component parts of the catalog.
+ *
+ * The touch points where you can replace or extend behavior are as follows:
+ *
+ * - Reader processors can be added or replaced. These implement the
+ *   functionality of reading raw data from a location, in the form of an
+ *   entity definition file.
+ * - Parser processors can be added or replaced. These accept the raw data as
+ *   read by the previous processors and parse it into raw structured data
+ *   (for example, from a binary buffer containing yaml text, to a JS object
+ *   structure).
+ * - Placeholder resolvers can be replaced or added. These run on the raw
+ *   structured data between the parsing and pre-processing steps, to replace
+ *   dollar-prefixed entries with their actual values (like $file).
+ * - Pre-processors can be added or replaced. These take the raw unvalidated
+ *   data from the parser processors and can enrich or extend it before
+ *   validation. This is the place where for example codeowners data can be
+ *   injected into partial entity definitions.
+ * - Entity policies can be added or replaced. These are the first line of
+ *   validation from the output of the pre-processing step. All policies are
+ *   given the chance to inspect the entity, and all of them have to pass in
+ *   order for the entity to be considered valid from an overall point of
+ *   view.
+ * - Field format validators can be replaced. These check the format of
```

Comment:
I think it may be easier to build this through things that are together, rather than through a framework where things are replaced. Thinking composition of exported arrays of e.g. `[...defaultPolicies.formatValidators()]`. This or maybe a bit of a combination between the two?

I'm thinking this leans very far towards inserting specific things into specific slots, which is inherently less flexible and difficult to maintain over time.